impl/src/main/java/io/kroxylicious/testing/kafka/common/KafkaTopology.java

```diff
@@ -0,0 +1,237 @@
+        kafkaTopology.nodeConfigurations.putAll(configs);
+        // createNode requires the topology to have all nodes configurations installed first, so it can ask for all
+        // controller node details.
+        Map<Integer, KafkaNode> nodes = configs.values().stream()
+                .collect(Collectors.toMap(KafkaNodeConfiguration::nodeId, driver::createNode));
+        kafkaTopology.addNodes(nodes);
+        return kafkaTopology;
+    }
+
+    @NotNull
+    private static Map<Integer, KafkaNodeConfiguration> generateConfigurations(KafkaDriver driver, KafkaClusterConfig config, KafkaTopology kafkaTopology) {
+        int nodesToCreate = Math.max(config.getBrokersNum(), config.getKraftControllers());
+        Stream<KafkaNodeConfiguration> nodeConfigurationStream = IntStream.range(0, nodesToCreate)
+                .mapToObj(value -> new IdAndRoles(roles(config, value), value))
+                // note that generation of a configuration may allocate resources like ports to it
+                .map(idAndRoles -> getNodeConfiguration(kafkaTopology, config, idAndRoles, driver));
+        return nodeConfigurationStream.collect(Collectors.toMap(KafkaNodeConfiguration::nodeId, configuration -> configuration));
+    }
+
+    private void addNodes(Map<Integer, KafkaNode> nodes) {
+        this.nodes.putAll(nodes);
+    }
+
+    public String getQuorumVoters() {
+        String collect = nodeDescriptions().stream().map(KafkaNodeConfiguration::controllerQuorumAddress)
+                .filter(Optional::isPresent)
+                .map(Optional::get)
+                .collect(Collectors.joining(","));
+        if (collect.isBlank()) {
+            throw new IllegalStateException("quorum voters empty! we should have some controllers in the cluster");
+        }
+        return collect;
+    }
+
+    public void close() {
+        if (this.zookeeper != null) {
+            this.zookeeper.close();
+        }
+    }
+
+    public void startZookeeperIfRequired() {
+        if (zookeeper != null) {
+            zookeeper.start();
+        }
+    }
+
+    public record IdAndRoles(Set<Role> roles, int nodeId) {
+    }
+
+    @NotNull
+    private static KafkaNodeConfiguration getNodeConfiguration(KafkaTopology kafkaTopology, KafkaClusterConfig config, IdAndRoles idAndRoles, KafkaDriver driver) {
+        return new KafkaNodeConfiguration(kafkaTopology, idAndRoles.nodeId, idAndRoles.roles, config, driver);
+    }
+
+    private static Set<Role> roles(KafkaClusterConfig config, int nodeId) {
+        if (!config.isKraftMode()) {
+            return Set.of(Role.BROKER);
+        }
+        else if (nodeId < config.getBrokersNum() && nodeId < config.getKraftControllers()) {
+            return Set.of(Role.BROKER, Role.CONTROLLER);
+        }
+        else if (nodeId < config.getBrokersNum()) {
+            return Set.of(Role.BROKER);
+        }
+        else {
+            return Set.of(Role.CONTROLLER);
+        }
```

Comment:
nit: I think BROKER should be the default case not Controller.

```suggestion
        else if (nodeId < config.getKraftControllers()) {
            return Set.of(Role.CONTROLLER);
        }
        else {
            return Set.of(Role.BROKER);
        }
```